### PR TITLE
[Time] Conductors and API Enhancements 

### LINKED ISSUE #4975: [Plots] Reduce space used by independent Time Conductors
**Is your feature request related to a problem? Please describe.**
We should use less vertical space for displaying independent time conductors (and the toggle button) in plots.

MORE DETAIL NEEDED

### LINKED ISSUE #5773: Global clock
**Is your feature request related to a problem? Please describe.**
Open MCT should provide a clock that is always ticking, irrespective of time conductor mode, to provide a canonical time value for "now". Many plugins need to know what the current time is according to the configured time system and clock, even if the time conductor is in "fixed" mode.

eg. Displays such as the time list view which display a moving "now" line relative to scheduled activities in the past and future.

`Date.now()` cannot be used for this purpose because Open MCT must support simulated times in the past and future, as well as non-terrestrial time systems.

**Describe the solution you'd like**
#### Stage 1: Separate the selection of "mode" from the selection of the "clock"
1. Modify the time conductor UI to add a "mode" drop-down that selects between fixed and real-time modes. A drop-down for selecting the clock will be maintained, however the "fixed" option will be removed from it, as this is now selected via the "mode". 
2. Selecting "real-time" mode from the drop-down will set the active clock to the one that is currently selected in the "clock" drop-down. Selecting "fixed" mode will unset it. The conductor is already doing this, but with a single drop-down instead.

#### Stage 2: Add API support for a canonical Open MCT clock
1. Modify the Time API to add a `mode()` getter/setter to time contexts. Setting the `mode` to `real-time` will start the configured clock, and cause the conductor's time context to automatically follow that clock, automatically updating the bounds as it currently does when following a clock. Setting the `mode` to `fixed` ~will stop the currently configured clock~.
2. Modify the Time API so that when a clock is set via the `clock()` setter that clock is also immediately started. The time context will only follow the clock if the `mode` is set to "real-time".

#### Stage 4: Modify Notebook entry timestamps to use the Open MCT clock instead of the wall clock.

** Open Questions**

* What about object timestamps? I suspect we still want to use `Date.now()` for this. Although it might cause some user confusion, we need to know the wall-clock time that objects were created and modified for various API-internal reasons.

### PR DESCRIPTION
**Pull request by jvigliotta**
Closes #4975 
Closes #5773

### Describe your changes:
- mode functionality added to TimeAPI
- TimeAPI modified to always have a ticking clock
- mode dropdown added to independent and regular time conductors
- overall conductor appearance modifications and enhancements
- TimeAPI methods deprecated with warnings
- Significant updates to markup, styling and behavior of main Time Conductor and independent version.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [ ] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Comment by charlesh88** [comment]
Todos as of 7/7/2023:
@jvigliotta 
- [ ] The read-only version of the current mode isn't updating on user selection from the pop-up:
Screen Shot 2023-07-07 at 10 45 00 AM
- [ ] Panning the fixed time ticks invokes the TC popup, but shouldn't.
- [ ] Regression: the fixed time ticks area doesn't allow brush zooming anymore but should.

@charlesh88 and maybe @jvigliotta 
- [ ] Size of the mini super-menu for mode and clock selection is messed up and needs to be fixed. Initial placement is good, but the menu expands size on mouseover and shouldn't.

**Comment by shefalijoshi** [comment]
Listing found issues here:

- [ ] There appear to be some UI issues around the Clock sub-menu. See attached recording: 
https://github.com/nasa/openmct/assets/5247066/3ca96bb8-4ae0-4d1e-a062-0c3e34ec7ed3

- [ ] The mode menu has some text bleeding through the menu area
image

- [ ] Switching from Real time mode to Fixed timespan does not change the time conductor bar's text to `FIXED TIMESPAN`. It continues to say `REAL-TIME`. Same for switching from Fixed to Realtime.
image

- [ ] After a refresh, clicking on the check mark icon in the time conductor popup results in a console error:
```
Uncaught TypeError: this.conductorPopup is null
    positionBox conductorPopUpManager.js:53
    n duration.js:44
    n duration.js:72
    initializePopup conductorPopUpManager.js:41
    VueJS 5
    numberSelect timePopupRealtime.vue:228
    setOffsets timePopupRealtime.vue:226
    mounted timePopupRealtime.vue:146
    VueJS 19
    showPopup conductorPopUpManager.js:43
    mounted conductorPopUpManager.js:31
```
See attached recording: 
https://github.com/nasa/openmct/assets/5247066/8f5e0a32-4163-4145-9656-8c3bc3b2d509

- [ ] Only the current clock set in the global time conductor is available in Independent time conductor clock sub menu. The others are not listed.
image

File: src/plugins/timeConductor/ConductorClock.vue [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    <div class="c-compact-tc__clock__value">{{ selectedClock.name }}</div>
+  </div>
+</template>
+
+<script>
+import modeMixin from './mode-mixin';
+import { TIME_CONTEXT_EVENTS } from '../../api/time/constants';
+
+export default {
+  mixins: [modeMixin],
```
Comment (shefalijoshi) [inline src/plugins/timeConductor/ConductorClock.vue:36]: Hmm. It doesn't look like there is much overlap between the clock menu and the mode menu. Could we separate the mode-mixin into two mixins? One for clocks and one for modes?

File: src/ui/components/ObjectFrame.vue [hunk 2]
```diff
@@ -144,6 +149,11 @@ export default {
   computed: {
     statusClass() {
       return this.status ? `is-status--${this.status}` : '';
+    },
+    supportsIndependentTime() {
+      // const viewKey = this.getViewKey();
+
+      return true; //this.domainObject && SupportedViewTypes.includes(viewKey);
```
Comment (shefalijoshi) [inline src/ui/components/ObjectFrame.vue:156]: Probably switch back to checking the viewKey? :)

File: src/ui/components/ObjectView.vue [hunk 3]
```diff
@@ -21,17 +21,12 @@
 -->
 <template>
   <div>
-    <div
-      v-if="supportsIndependentTime"
+    <!-- <div
```
Comment (shefalijoshi) [inline src/ui/components/ObjectView.vue:24]: Let's remove this code if it's no longer needed.

File: src/ui/components/ObjectView.vue [hunk 4]
```diff
@@ -40,19 +35,19 @@
...
-  'scatter-plot.view',
-  'time-strip.view'
-];
+// const SupportedViewTypes = [
+//   'plot-stacked',
+//   'plot-overlay',
+//   'bar-graph.view',
+//   'scatter-plot.view',
+//   'time-strip.view'
+// ];
```
Comment (shefalijoshi) [inline src/ui/components/ObjectView.vue:47]: Please remove.

File: src/ui/components/ObjectView.vue [hunk 5]
```diff
@@ -99,11 +94,11 @@ export default {
     font() {
       return this.objectFontStyle ? this.objectFontStyle.font : this.layoutFont;
     },
-    supportsIndependentTime() {
-      const viewKey = this.getViewKey();
+    // supportsIndependentTime() {
```
Comment (shefalijoshi) [inline src/ui/components/ObjectView.vue:97]: Also remove :)

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
This is a LOT of hard work. Thank you!

A few changes requested and found some issues while testing.

**Comment by jvigliotta** [comment]
> Todos as of 7/7/2023: @jvigliotta
> 
> * [ ]  The read-only version of the current mode isn't updating on user selection from the pop-up:
>   Screen Shot 2023-07-07 at 10 45 00 AM
> * [x]  Panning the fixed time ticks invokes the TC popup, but shouldn't.
> * [ ]  Regression: the fixed time ticks area doesn't allow brush zooming anymore but should.
> 
> @charlesh88 and maybe @jvigliotta
> 
> * [ ]  Size of the mini super-menu for mode and clock selection is messed up and needs to be fixed. Initial placement is good, but the menu expands size on mouseover and shouldn't.

@charlesh88 going through these. The zoom was actually working, but the classes weren't updated to work with the new compact mode class, I've moved those styles over now and seems to be good for both zooming and panning styling.

File: src/api/time/TimeAPI.js [hunk 6]
```diff
@@ -180,11 +183,12 @@ class TimeAPI extends GlobalTimeContext {
       objectPath.length && this.openmct.objects.makeKeyString(objectPath[0].identifier);
 
     if (!viewKey) {
-      // Return the global time context
+      // Return the global time contex
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/api/time/TimeAPI.js:186]: yep!

File: src/plugins/timeConductor/ConductorClock.vue (hunk 1 above)
Comment (jvigliotta) [reply src/plugins/timeConductor/ConductorClock.vue:36]: Yep! Definitely. They diverged quite a bit... not any overlap I think :)

File: src/plugins/plot/MctPlot.vue [hunk 7]
```diff
@@ -521,20 +521,19 @@ export default {
...
+      this.updateDisplayBounds(this.timeContext.getBounds());
+      this.timeContext.on('modeChanged', this.updateMode);
+      this.timeContext.on('boundsChanged', this.updateDisplayBounds);
       this.synchronized(true);
     },
     stopFollowingTimeContext() {
       if (this.timeContext) {
-        this.timeContext.off('clock', this.updateRealTime);
-        this.timeContext.off('bounds', this.updateDisplayBounds);
+        this.timeContext.off('clockChanged', this.updateMode);
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/plot/MctPlot.vue:535]: haha, yeah, not sure what was going on there. Thanks!

File: src/plugins/URLTimeSettingsSynchronizer/URLTimeSettingsSynchronizer.js [hunk 8]
```diff
@@ -218,18 +220,17 @@ export default class URLTimeSettingsSynchronizer {
...
 
-    if (isValid) {
-      if (mode.toLowerCase() === MODE_FIXED) {
-        isValid = true;
-      } else {
-        isValid = this.openmct.time.clocks.get(mode) !== undefined;
-      }
+    if (
+      isValid &&
+      (mode.toLowerCase() === FIXED_MODE_KEY || mode.toLowerCase() === REALTIME_MODE_KEY)
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/URLTimeSettingsSynchronizer/URLTimeSettingsSynchronizer.js:230]: That's a great point. I may need to add the clock check back in.

[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/URLTimeSettingsSynchronizer/URLTimeSettingsSynchronizer.js:230]: So... it definitely was not correct. I think I was thinking it was a realtime check, but that's not the case here. I've reverted!

File: src/ui/components/ObjectFrame.vue (hunk 2 above)
Comment (jvigliotta) [reply src/ui/components/ObjectFrame.vue:156]: gotcha!

File: src/ui/components/ObjectView.vue (hunk 5 above)
Comment (jvigliotta) [reply src/ui/components/ObjectView.vue:97]: got it!

File: src/ui/components/ObjectView.vue (hunk 4 above)
Comment (jvigliotta) [reply src/ui/components/ObjectView.vue:47]: done!

File: src/ui/components/ObjectView.vue (hunk 3 above)
Comment (jvigliotta) [reply src/ui/components/ObjectView.vue:24]: done!

**Comment by jvigliotta** [comment]
@charlesh88 @shefalijoshi I'm not really sure what's up with the menu styles being all jacked. I'll dig in a little further, but I think this is ready for another look.

**Comment by charlesh88** [comment]
@jvigliotta 
- Just pulled and merged with my (extensive) local changes. Looking very good; ticks panning and brushing is working  great - thanks for the fix there. I'll push my stuff shortly.
- Not sure either what's going on with the menus. I have a vague recollection of fixing all that in the branch that Shefali and I were working in, but I'm not sure. I'm going to go back and have a look, but even if I haven't I don't think it's going to be a big deal to fix now.

I added a new todo for you in my list of issues above, now dated "as of 7/11/2023" - please have a look.

File: src/api/time/TimeContext.js [hunk 9]
```diff
@@ -353,25 +365,298 @@ class TimeContext extends EventEmitter {
...
+     * @event bounds
+     * @memberof module:openmct.TimeAPI~
+     * @property {TimeConductorBounds} bounds The newly updated bounds
+     * @property {boolean} [tick] `true` if the bounds update was due to
+     * a "tick" event (i.e. was an automatic update), false otherwise.
+     */
+    this.emit(TIME_CONTEXT_EVENTS.boundsChanged, this.boundsVal, false);
+
+    //Return a copy to prevent direct mutation of time conductor bounds.
+    return this.#copy(this.boundsVal);
```
Comment (akhenry) [inline src/api/time/TimeContext.js:488]: Why does `setBounds()` return anything? Moreover, why does it take a _second_ copy of the bounds object?

File: src/api/time/TimeContext.js [hunk 10]
```diff
@@ -353,25 +365,298 @@ class TimeContext extends EventEmitter {
...
+     * @property {TimeSystem} The value of the currently applied
+     * Time System
+     * */
+    this.emit(TIME_CONTEXT_EVENTS.timeSystemChanged, this.#copy(this.system));
+
+    if (bounds) {
+      this.setBounds(bounds);
+    }
+
+    return this.system;
```
Comment (akhenry) [inline src/api/time/TimeContext.js:443]: Why are we returning anything from setter functions?

**Review by akhenry — CHANGES_REQUESTED** [review]
API changes look

File: src/api/time/TimeContext.js (hunk 9 above)
Comment (jvigliotta) [reply src/api/time/TimeContext.js:488]: you know Andrew... that's a great question. I will remove it :)

File: src/api/time/TimeContext.js [hunk 11]
```diff
@@ -353,25 +365,298 @@ class TimeContext extends EventEmitter {
...
+   * @param {TimeSystem | string} timeSystemOrKey
+   * @param {module:openmct.TimeAPI~TimeConductorBounds} bounds
+   * @fires module:openmct.TimeAPI~timeSystem
+   * @returns {TimeSystem} The currently applied time system
+   * @memberof module:openmct.TimeAPI#
+   * @method setTimeSystem
+   */
+  setTimeSystem(timeSystemOrKey, bounds) {
+    if (!this.isRealTime() && !bounds) {
+      throw new Error('Must specify bounds when changing time system without an active clock.');
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/api/time/TimeContext.js:402]: yep! Will update.

File: src/api/time/TimeContext.js (hunk 10 above)
Comment (jvigliotta) [reply src/api/time/TimeContext.js:443]: Will update!

File: src/api/time/IndependentTimeContext.js [hunk 12]
```diff
@@ -46,35 +47,59 @@ class IndependentTimeContext extends TimeContext {
...
+  bounds() {
     if (this.upstreamTimeContext) {
       return this.upstreamTimeContext.bounds(...arguments);
     } else {
       return super.bounds(...arguments);
     }
   }
 
-  tick(timestamp) {
+  getBounds() {
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/api/time/IndependentTimeContext.js:58]: Have filed a followup - https://github.com/nasa/openmct/issues/6799

File: src/plugins/clock/components/ClockIndicator.vue [hunk 13]
```diff
@@ -46,16 +46,17 @@ export default {
     };
   },
   mounted() {
-    this.unlisten = ticker.listen(this.tick);
+    this.tick = raf(this.tick);
```
Comment (shefalijoshi) [inline src/plugins/clock/components/ClockIndicator.vue:49]: In Remote-Clock mode the indicator updates the time very slowly - once every few seconds.
Is this expected?

File: src/plugins/URLTimeSettingsSynchronizer/URLTimeSettingsSynchronizer.js [hunk 14]
```diff
@@ -106,30 +109,32 @@ export default class URLTimeSettingsSynchronizer {
...
 
   setTimeApiFromUrl(timeParameters) {
-    if (timeParameters.mode === 'fixed') {
-      if (this.openmct.time.timeSystem().key !== timeParameters.timeSystem) {
-        this.openmct.time.timeSystem(timeParameters.timeSystem, timeParameters.bounds);
-      } else if (!this.areStartAndEndEqual(this.openmct.time.bounds(), timeParameters.bounds)) {
-        this.openmct.time.bounds(timeParameters.bounds);
+    const timeSystem = this.openmct.time.getTimeSystem();
+
+    if (timeParameters.mode === FIXED_MODE_KEY) {
```
Comment (shefalijoshi) [inline src/plugins/URLTimeSettingsSynchronizer/URLTimeSettingsSynchronizer.js:114]: We have no way of setting the clock anymore. Perhaps create a new issue for that?

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
Some changes requested. And some questions

**Review by shefalijoshi — APPROVED** [review]

**Pull request merged**